### PR TITLE
Dbaas backups

### PR DESCRIPTION
**Pull request by schirevko**
Add vkcs_db_backup resource
Add restore_point field to dbaas instances and clusters

### CODE REVIEW

File: vkcs/db_backup.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	"github.com/gophercloud/gophercloud"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func databaseBackupStateRefreshFunc(client databaseClient, backupID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		b, err := dbBackupGet(client, backupID).extract()
+		if err != nil {
+			if _, ok := err.(gophercloud.ErrDefault404); ok {
+				return b, "DELETED", nil
```
Comment (ftersin) [inline vkcs/db_backup.go:15]: could we use a named constant here?

File: vkcs/resource_vkcs_db_cluster.go [hunk 2]
```diff
@@ -269,6 +269,26 @@ func resourceDatabaseCluster() *schema.Resource {
...
 				},
 			},
+			"restore_point": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"backup_ref": {
```
Comment (ftersin) [inline vkcs/resource_vkcs_db_cluster.go:279]: probably "backup_id" is better

File: vkcs/db_backup.go (hunk 1 above)
Comment (schirevko) [reply vkcs/db_backup.go:15]: I think we could, but this style (without named constants) is used everywhere in similar methods of provider, thus we will need to replace code in many places to maintain similarity.

Comment (ftersin) [reply vkcs/db_backup.go:15]: ok

File: vkcs/resource_vkcs_db_cluster.go (hunk 2 above)
Comment (schirevko) [reply vkcs/resource_vkcs_db_cluster.go]: Renamed

File: vkcs/db_instance_client.go [hunk 3]
```diff
@@ -570,6 +621,26 @@ func clusterGetCapabilities(client databaseClient, id string) (r getInstanceCapa
...
+	reqOpts := getDBRequestOpts(200)
+	var result *http.Response
+	result, r.Err = client.Get(backupScheduleURL(client, instancesAPIPath, id), &r.Body, reqOpts)
+	if r.Err == nil {
+		r.Header = result.Header
+	}
+	return
+}
+
+func clusterGetBackupSchedule(client databaseClient, id string) (r getClusterBackupScheduleResult) {
```
Comment (ftersin) [inline vkcs/db_instance_client.go:634]: should not we need to move this to cluster client?

**Review by ftersin — CHANGES_REQUESTED** [review]
also we need full examples for on-demand and pitr bakups

File: docs/resources/vkcs_db_cluster_with_shards.md [hunk 4]
```diff
@@ -88,6 +88,10 @@ resource "vkcs_db_cluster_with_shards" "db-cluster-with-shards" {
         * `uuid` - The id of the network. Changing this creates a new cluster.
         * `port` - The port id of the network. Changing this creates a new cluster.
 
+* `restore_point` - Object that represents backup to restore instance from. It has following attributes:
+    * `backup_id` - (Required) ID of the backup.
+    * `target` - Timestamp of needed backup. You can specify "latest" to use most recent backup.
```
Comment (ftersin) [inline docs/resources/vkcs_db_cluster_with_shards.md:93]: no one datastore with shards supports pitr backups

File: docs/resources/vkcs_db_cluster.md [hunk 5]
```diff
@@ -83,6 +83,18 @@ The following arguments are supported:
     * `name` - (Required) The name of the capability to apply.
     * `settings` - Map of key-value settings of the capability.
 
+* `restore_point` - Object that represents backup to restore cluster from. It has following attributes:
+    * `backup_id` - (Required) ID of the backup.
+    * `target` - Timestamp of needed backup. You can specify "latest" to use most recent backup.
```
Comment (ftersin) [inline docs/resources/vkcs_db_cluster.md:88]: probably we must mention here that this works with pitr backups only

File: docs/data_sources/vkcs_db_backup.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+The following attributes are exported:
+
+* `name` - The name of the backup. Changing this creates a new backup
+
+* `dbms_id` - ID of the instance or cluster, to create backup of.
+
+* `description` - The description of the backup
+
+* `container_prefix` - ??
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/data_sources/vkcs_db_backup.md]: Added

File: docs/data_sources/vkcs_db_backup.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+* `location_ref` - ??
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/data_sources/vkcs_db_backup.md]: Added

File: docs/resources/vkcs_db_backup.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+## Example Usage
+
+```terraform
+
+resource "vkcs_db_instance" "db-instance" {
+  name = "db-instance"
+
+  datastore {
+    type    = example_datastore_type
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/resources/vkcs_db_backup.md:21]: This value is used in instance resource block, should we change it in whole doc, not only in backup?

File: docs/resources/vkcs_db_cluster.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/resources/vkcs_db_cluster.md]: Added examples

File: docs/resources/vkcs_db_cluster.md [hunk 9]
```diff
@@ -83,6 +83,18 @@ The following arguments are supported:
     * `name` - (Required) The name of the capability to apply.
     * `settings` - Map of key-value settings of the capability.
 
+* `restore_point` - Object that represents backup to restore cluster from. It has following attributes:
+    * `backup_id` - (Required) ID of the backup.
+    * `target` - Timestamp of needed backup. You can specify "latest" to use most recent backup.
+
+* `backup_schedule` - Object that represents configuration of PITR backup. This functionality is available only for postgres datastore. This object has following attributes:
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/resources/vkcs_db_cluster.md:148]: Added examples

File: vkcs/db_instance_client.go (hunk 3 above)
Comment (schirevko) [reply vkcs/db_instance_client.go]: Moved

File: docs/resources/vkcs_db_cluster.md (hunk 5 above)
Comment (schirevko) [reply docs/resources/vkcs_db_cluster.md]: I think, it should work with all backups, doesn't it? We can create regular backup and specify its id here to restore dbms from it

File: docs/data_sources/vkcs_db_backup.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+The following attributes are exported:
+
+* `name` - The name of the backup. Changing this creates a new backup
+
+* `dbms_id` - ID of the instance or cluster, to create backup of
+
+* `description` - The description of the backup
+
+* `container_prefix` - Prefix of S3 bucket (-<project_id>) to store backup data. Default: databasebackups
```
Comment (ftersin) [inline docs/data_sources/vkcs_db_backup.md:37]: imposibble to guess what is included in parentheses

File: docs/resources/vkcs_db_backup.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ftersin) [reply docs/resources/vkcs_db_backup.md:21]: not everywhere

File: examples/create-vkcs-db-backup/main.tf [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    settings = {
+      "listen_port" : "9100"
+    }
+  }
+}
+
+resource "vkcs_db_backup" "db-backup" {
+  name = "db-backup"
+  dbms_id = vkcs_db_instance.db-instance.id
+}
```
Comment (ftersin) [inline examples/create-vkcs-db-backup/main.tf:58]: add empty line here and below

File: docs/resources/vkcs_db_cluster.md [hunk 12]
```diff
@@ -32,9 +55,43 @@ resource "vkcs_db_cluster" "mydb-cluster" {
...
+
+  volume_size = 10
+  volume_type = example_volume_type
+
+  network {
+    uuid = example_network_id
+  }
+
+  backup_schedule {
+    name = backup_schedule_name
```
Comment (ftersin) [inline docs/resources/vkcs_db_cluster.md:87]: rename the name to smth more real pls

File: docs/resources/vkcs_db_instance.md [hunk 13]
```diff
@@ -98,6 +184,17 @@ The following arguments are supported:
     * `name` - (Required) The name of the capability to apply.
     * `settings` - Map of key-value settings of the capability.
 
+* `restore_point` - Object that represents backup to restore instance from. It has following attributes:
```
Comment (ftersin) [inline docs/resources/vkcs_db_instance.md:187]: don't we need to add smth of these to instance/cluster datasource?

File: examples/create-vkcs-db-backup/main.tf [hunk 14]
```diff
@@ -0,0 +1,58 @@
+terraform {
+  required_providers {
+    vkcs = {
+      source  = "vk-cs/vkcs"
+      version = "~> 0.1.0"
```
[the remark this answers is not in the document]
Comment (schirevko) [reply examples/create-vkcs-db-backup/main.tf]: Changed

File: docs/resources/vkcs_db_cluster.md (hunk 12 above)
Comment (schirevko) [reply docs/resources/vkcs_db_cluster.md]: Done

File: docs/data_sources/vkcs_db_backup.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
+---
+layout: "vkcs"
+page_title: "vkcs: db_backup"
+subcategory: ""
+description: |-
+  Get information on a db backup.
+---
+
```
[the remark this answers is not in the document]
Comment (schirevko) [reply docs/data_sources/vkcs_db_backup.md]: Done

File: docs/resources/vkcs_db_instance.md (hunk 13 above)
Comment (ftersin) [reply docs/resources/vkcs_db_instance.md]: we don't have such datasources

**Review by pashapdev — APPROVED** [review]

**Pull request merged**